### PR TITLE
Add `eval` spicepod component; basic HTTP api to run eval. 

### PR DESCRIPTION
**Pull request by Jeadie**
# 🗣 Description
## `eval` spicepod component
```yaml
evals:
  - name: australia
    description: Make sure the model is a true blue Aussie
    dataset: cricket_logic # This is just any compatible dataset component
    scorers:
      - Match # This is a builtin scorer; model graded scorer support coming!
      
datasets:
  - name: cricket_logic
    from: file:samples.jsonl
```

## HTTP api `/v1/eval/:name/run`
 - This is a temporary API, and UX will change significantly. For one, it directly runs the eval, forcing the client to wait until completion. 
```shell
curl -XPOST "http://localhost:8090/v1/eval/australia/run" \
  -d '{
    "model": "oai"
  }'
```

```json
{
    "Match": [
        [
            "mean",
            0.0
        ],
        [
            "std_dev",
            0.0
        ]
    ]
}
```

## 🔨 Related Issues
 - #3320 

## 📄 Documentation Updates
- [ ] `eval` spicepod component
- [ ] builtin eval scorers

### CODE REVIEW

File: crates/runtime/src/model/eval_scorer.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    fn metrics(&self, scores: &[f32]) -> Vec<(String, f32)>;
+}
+
+/// [`MatchScorer`] checks for equality between the actual and ideal values.
+///
+/// The score is an exact match, but it only compare the less detailed of the two [`DatasetOutput`]. ([`DatasetOutput::AssistantResponse`] < [`DatasetOutput::Messages`]).
+pub struct MatchScorer {}
+
+#[async_trait]
+impl Scorer for MatchScorer {
```
Comment (sgrebnov) [inline crates/runtime/src/model/eval_scorer.rs:41]: Please see if it makes sense to add unit tests

**Review by sgrebnov — DISMISSED** [review]

File: crates/runtime/src/model/eval.rs [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+use llms::chat::Chat;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use snafu::{ResultExt, Snafu};
+use spicepod::component::eval::Eval;
+use std::{collections::HashMap, sync::Arc};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to query eval dataset '{dataset_name}': {source}"))]
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval.rs]: Is there any action a user can take for any of these errors?

File: crates/runtime/src/model/eval.rs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+        Ok(result)
+    }
+}
+
+#[allow(clippy::borrowed_box, clippy::implicit_hasher)]
+pub async fn run_eval(
+    eval: &Eval,
+    df: Arc<DataFusion>,
+    model: &Box<dyn Chat>,
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval.rs]:
```suggestion
    model: &dyn Chat,
```

Pay attention to lints

File: crates/runtime/src/model/eval.rs [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+                    return Err(Error::InvalidOutputFormat { reason: format!("could not convert JSON format of an element into a known model output format: {e}") });
+                }
+            };
+        }
+
+        Ok(result)
+    }
+}
+
+#[allow(clippy::borrowed_box, clippy::implicit_hasher)]
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval.rs]:
```suggestion
#[allow(clippy::implicit_hasher)]
```

File: crates/runtime/src/model/eval.rs [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+        let Some(scorer) = scorers.get(name) else {
+            return Err(Error::EvalScorerUnavailable {
+                scorer_name: name.clone(),
+                eval_name: eval_name.clone(),
+            });
+        };
+        scorers_subset.insert(name, scorer);
+    }
+
+    let dataset =
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval.rs:329]: Does this need to be sanitized?

File: crates/runtime/src/model/eval_scorer.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            ("mean".to_string(), mean),
+            // For  Bernoulli r.v., the variance is p(1-p).
+            ("std_dev".to_string(), (mean * (1.0 - mean)).sqrt()),
+        ]
+    }
+}
+
+#[must_use]
+pub fn builtin_scorer() -> Vec<(&'static str, Arc<dyn Scorer>)> {
+    vec![("Match", Arc::new(MatchScorer {}))]
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval_scorer.rs]: Should this be upper case? Most of our configuration is lower case

File: crates/runtime/src/http/routes.rs [hunk 7]
```diff
@@ -75,6 +75,7 @@ pub(crate) fn routes(
             .route("/v1/search", post(v1::search::post))
             .route("/v1/tools", get(v1::tools::list))
             .route("/v1/tool/:name", post(v1::tools::post))
+            .route("/v1/eval/:name/run", post(v1::eval::post))
```
Comment (lukekim) [inline crates/runtime/src/http/routes.rs]:
```suggestion
            .route("/v1/evals/:name", post(v1::eval::post))
```

You don't need the `/run` a POST on the object is sufficient.  Or if this is truly a "eval run", then it should be `POST /v1/evals/:name/runs`

I also noticed that tools is incorrect. It should be plural, `/tools/`

File: crates/runtime/src/model/eval.rs (hunk 2 above)
Comment (Jeadie) [reply crates/runtime/src/model/eval.rs]: Yes, I should add.

File: crates/runtime/src/model/eval.rs [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+    scorers: &HashMap<String, Arc<dyn Scorer>>,
+) -> Result<HashMap<String, Vec<(String, f32)>>> {
+    let Eval {
+        name: eval_name,
+        scorers: scorer_names,
+        dataset: dataset_str,
+        ..
+    } = eval;
+
+    let mut scorers_subset = HashMap::with_capacity(scorer_names.len());
```
Comment (Sevenannn) [inline crates/runtime/src/model/eval.rs:313]: Just curious about the variable naming here, why is it called scorers_subset, it seems to contain all the scores used by eval

Comment (Jeadie) [reply crates/runtime/src/model/eval.rs:313]: Not amazing naming, `scorers` is all available scorers (spiced-wide). `scorers_subset` is the subset of all scorers needed for eval.

Comment (Jeadie) [reply crates/runtime/src/model/eval.rs:313]: This naming will go away in the followup PR #3780

File: crates/runtime/src/http/routes.rs [hunk 9]
```diff
@@ -74,7 +74,9 @@ pub(crate) fn routes(
             .route("/v1/embeddings", post(v1::embeddings::post))
             .route("/v1/search", post(v1::search::post))
             .route("/v1/tools", get(v1::tools::list))
+            .route("/v1/tools/:name", post(v1::tools::post))
             .route("/v1/tool/:name", post(v1::tools::post))
```
Comment (phillipleblanc) [inline crates/runtime/src/http/routes.rs:79]: We should add a comment this route is deprecated (so future people that come across this don't wonder why we have two, and which one is the correct one)

File: crates/runtime/src/model/eval.rs [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+    }
+
+    Ok(scorers_subset
+        .iter()
+        .map(|(name, scorer)| ((*name).clone(), scorer.metrics(&aggregate[*name])))
+        .collect())
+}
+
+/// Return format of [`DatasetOutput`] determined by `output_format`. `output_format` can be empty, is only used for its enum type.
+#[allow(clippy::borrowed_box)]
```
Comment (phillipleblanc) [inline crates/runtime/src/model/eval.rs]:
```suggestion
```

**Review by phillipleblanc — DISMISSED** [review]

**Review by sgrebnov — APPROVED** [review]

**Pull request merged**